packages/router/src/lib/operations/metaTx.ts

```diff
@@ -0,0 +1,267 @@
+  logger.debug("Method start", requestContext, methodContext, { input });
+
+  // Validate Input schema
+  const validateInput = ajv.compile(MetaTxInputSchema);
+  const validInput = validateInput(input);
+  if (!validInput) {
+    const msg = validateInput.errors?.map((err: any) => `${err.instancePath} - ${err.message}`).join(",");
+    throw new ParamsInvalid({
+      methodContext,
+      requestContext,
+      paramsError: msg,
+    });
+  }
+
+  const { chainId, data, type, to } = input;
+
+  let receipt;
+
+  if (!config.chainConfig[chainId]) {
+    throw new NoChainConfig(chainId, { methodContext, requestContext, input });
+  }
+
+  const relayerFeeLowerBound = config.chainConfig[chainId].relayerFeeThreshold;
+  if (!config.allowRelay && !config.chainConfig[chainId].allowRelay) {
+    // TODO: should we throwing error if relayer is not allowed?
+    // router shouldn't listen to metaTx request in first place if it's not allowed to relay
+    throw new NotAllowedFulfillRelay(chainId, {
+      methodContext,
+      requestContext,
+      data,
+      input,
+    });
+  }
+
+  // receiver fulfill, done directly on contract using user's sig broadcast
+  if (type === MetaTxTypes.Fulfill) {
+    const { txData, signature, relayerFee, callData } = data as MetaTxFulfillPayload;
+    // Send to tx service
+    logger.info("Sending fulfill tx", requestContext, methodContext, { signature });
+
+    let outputDecimals = chainData.get(txData.receivingChainId.toString())?.assetId[txData.receivingAssetId]?.decimals;
+    if (!outputDecimals) {
+      outputDecimals = await txService.getDecimalsForAsset(txData.receivingChainId, txData.receivingAssetId);
+    }
+    logger.info("Got output decimals", requestContext, methodContext, { outputDecimals });
+    const expectedFulfillFee = await calculateGasFeeInReceivingTokenForFulfill(
+      txData.receivingAssetId,
+      txData.receivingChainId,
+      outputDecimals,
+      requestContext,
+    );
+    logger.info("Expected Fulfill fee in router side", requestContext, methodContext, {
+      expectedFulfillFee: expectedFulfillFee.toString(),
+    });
+
+    const recvAmountLowerBound = expectedFulfillFee.mul(100 - relayerFeeLowerBound).div(100);
+    if (BigNumber.from(data.relayerFee).lt(recvAmountLowerBound)) {
+      throw new NotEnoughRelayerFee(chainId, {
+        methodContext,
+        requestContext,
+        relayerFee: data.relayerFee,
+        recvAmountLowerBound: recvAmountLowerBound.toString(),
+        input,
+        type,
+      });
+    }
+
+    receipt = await contractWriter.fulfillTransactionManager(
+      chainId,
+      {
+        txData,
+        signature: signature,
+        relayerFee: relayerFee,
+        callData: callData,
+      },
+      requestContext,
+    );
+
+    logger.info("Method complete", requestContext, methodContext, { transactionHash: receipt.transactionHash });
+  } else {
+    // router contract methods
```

Comment:
any router can also participate in relaying other routers contract txs? dont we want to offload all of this to gelato (that has proper fallbacks), or *only* relay if gelato fails

Reply:
Yes, it uses Gelato first then falls back here. This is the listening side. This was all quite complicated to architect and build lol.